pkg/natdiscovery/natdiscovery.go

```diff
@@ -0,0 +1,188 @@
+	return newNatDiscovery(localEndpoint)
+}
+
+func newNatDiscovery(localEndpoint *types.SubmarinerEndpoint) (*natDiscovery, error) {
+	port, err := extractNATDiscoveryPort(localEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	requestCounter, err := randomRequestCounter()
+	if err != nil {
+		return nil, err
+	}
+
+	return &natDiscovery{
+		localEndpoint:   localEndpoint,
+		serverPort:      port,
+		remoteEndpoints: map[string]*remoteEndpointNAT{},
+		findSrcIP:       findPreferredSourceIP,
+		requestCounter:  requestCounter,
+	}, nil
+}
+
+func randomRequestCounter() (uint64, error) {
+	n, err := rand.Int(rand.Reader, big.NewInt(1000))
+	if err != nil {
+		return 0, errors.Wrapf(err, "generatimg random request counter")
+	}
+
+	return n.Uint64(), nil
+}
+
+var errorNoNatDiscoveryPort = errors.New("no nat discovery port in endpoint")
+
+func extractNATDiscoveryPort(endpoint *types.SubmarinerEndpoint) (int, error) {
+	if endpoint.Spec.PortConfig == nil {
+		return 0, errorNoNatDiscoveryPort
+	}
+
+	if portString, ok := endpoint.Spec.PortConfig[v1.NATDiscoveryPortConfig]; ok {
+		port, err := strconv.Atoi(portString)
+		if err != nil {
+			return 0, errors.Wrapf(err, "error trying to extract port number from %q in %s",
+				portString, v1.NATDiscoveryPortConfig)
+		}
+
+		return port, nil
+	}
+
+	return 0, errorNoNatDiscoveryPort
+}
+
+func (nd *natDiscovery) Run(stopCh <-chan struct{}, natResultCh chan remoteEndpointNAT) error {
+	err := nd.runListener(stopCh)
+	if err != nil {
+		return err
+	}
+
+	go wait.Until(func() {
+		nd.checkEndpointList()
+	}, time.Second, stopCh)
+
+	return nil
+}
+
+func (nd *natDiscovery) AddEndpoint(endpoint *types.SubmarinerEndpoint) {
+	remoteNAT := newRemoteEndpointNAT(endpoint)
+
+	// support a remote cluster endpoint which still hasn't implemented this protocol
+	if _, err := extractNATDiscoveryPort(endpoint); err == errorNoNatDiscoveryPort {
+		remoteNAT.useLegacyNATSettings()
+	}
+
+	nd.Lock()
+	defer nd.Unlock()
+
+	nd.remoteEndpoints[endpoint.Spec.CableName] = remoteNAT
+}
+
+func (nd *natDiscovery) RemoveEndpoint(endpointName string) {
+	nd.Lock()
+	defer nd.Unlock()
+	delete(nd.remoteEndpoints, endpointName)
+}
+
+func (nd *natDiscovery) checkEndpointList() {
+	nd.Lock()
+	defer nd.Unlock()
+
+	for _, endpointNAT := range nd.remoteEndpoints {
+		if endpointNAT.hasTimedOut() {
+			endpointNAT.useLegacyNATSettings()
+		} else if endpointNAT.shouldCheck() {
+			err := nd.sendCheckRequest(endpointNAT)
+			if err != nil {
+				klog.Errorf("error sending check request to %q: %s", endpointNAT.Endpoint.Spec.CableName, err)
+			}
+		}
+	}
+}
+
+var UnknownEndpoint = errors.New("unknown endpoint")
+
+func (nd *natDiscovery) getRemoteEndpointNAT(endpointName string) *remoteEndpointNAT {
+	nd.Lock()
+	defer nd.Unlock()
+
+	remoteNat, ok := nd.remoteEndpoints[endpointName]
+	if ok {
+		return remoteNat
+	}
+
+	return nil
+}
+
+func (nd *natDiscovery) GetEndpointNAT(endpointName string) (bool, error) {
```

Comment:
```suggestion
func (nd *natDiscovery) UseNATForEndpoint(endpointName string) (bool, error) {
```

I haven't seen the usage of this API but it seems the usage would be:

```
if nd.UseNATForEndpoint(...) {
    nd.GetEndpointIP(...)
}
```

If so perhaps we should combine these methods, eg

`func (nd *natDiscovery) GetEndpointNATInfo(endpointName string) (useNat bool, ip string, bool found) `...

We could also add `error` although not necessary in this case.

Reply:
You're right, I'm removing this API. In the end I will be sending the ready endpoints via a channel.

Since it's a little bit asynchronous:

1) InstallCable will happen because we know of a new endpoint, at that moment we pass the endpoint to the natDiscovery controller,
2) Once the destination IP is discovered (or changed, public->private), it's passed to cableEngine again
3) Cable engine does the final installation.

Reply:
I removed both, and moving into this patch the "channel" implementation.

Reply:
ok - makes sense